### ISSUE TITLE
Pattern matching unifier normalizes too much

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#3578]
Pattern matching should preserve user-written types as far as possible, i.e., perform only to the minimal necessary weak head reductions.  This is important when doing proofs interactively.
```agda
open import Agda.Builtin.Nat

postulate
  MustNotBeNamed : Set

A = MustNotBeNamed

works : A → A
works x = {!!}

-- Goal: A
-- ——————————————
-- x : A

data D : Set where
 c : A → D

works' : D → A
works' (c x) = {!!}

-- Goal: A
-- ——————————————
-- x : A

data P : Nat → Set where
  z : A → P zero

test : ∀{n} → P n → A
test (z x) = {!!}

-- Goal: A
-- ——————————————
-- x : MustNotBeNamed
```
The unfolding of `A` in the `test` example is clearly unnecessary.

It seems to work as long as no unification is required.

**Comment by andreasabel** [comment]
I am afraid this regression was introduced by the new unifier released with 2.5.1.

**Comment by andreasabel** [comment]
Fixed by not keeping `UnifyState` in reduced form, rather reducing lazily when we need it.
